### PR TITLE
Add Thread Spinning Session Option in WinML

### PR DESCRIPTION
**Pull request by orilevari**
**Description**: Plumb through an API to WinML in SessionOptions to toggle the thread spinning Ort Session Option Config Entry. Includes tests as well which utilize a new adapter api to extract the internal config entry.

**Motivation and Context**
Allows clients to disable thread spinning in the case that they need to prioritize other CPU workloads.

### CODE REVIEW

File: winml/adapter/winml_adapter_apis.h [hunk 1]
```diff
@@ -56,6 +56,7 @@ ORT_API_STATUS(SessionRegisterGraphTransformers, _In_ OrtSession* session);
 ORT_API_STATUS(SessionRegisterCustomRegistry, _In_ OrtSession* session, _In_ IMLOperatorRegistry* registry);
 ORT_API_STATUS(SessionCopyOneInputAcrossDevices, _In_ OrtSession* session, _In_ const char* const input_name, _In_ OrtValue* orig_value, _Outptr_ OrtValue** new_value);
 ORT_API_STATUS(SessionGetNumberOfIntraOpThreads, _In_ OrtSession* session, _Out_ uint32_t* num_threads);
+ORT_API_STATUS(SessionGetIntraOpThreadSpinningConfigEntry, _In_ OrtSession* session, _Out_ const char** allow_spinning);
```
Comment (smk2007) [inline winml/adapter/winml_adapter_apis.h:59]: why char?

File: winml/api/windows.ai.machinelearning.native.idl [hunk 2]
```diff
@@ -35,10 +35,11 @@ interface ILearningModelDeviceFactoryNative : IUnknown
     HRESULT CreateFromD3D12CommandQueue(ID3D12CommandQueue * value, [out] IUnknown ** result);
 };
 
-[uuid(c71e953f-37b4-4564-8658-d8396866db0d), object, local]
+[uuid(5da37a26-0526-414b-91e4-2a0fa3ddba40), object, local]
 interface ILearningModelSessionOptionsNative : IUnknown
 {
     HRESULT SetIntraOpNumThreadsOverride(UINT32 intraOpNumThreads);
+    HRESULT SetIntraOpThreadSpinning(boolean allowSpinning);
```
Comment (smk2007) [inline winml/api/windows.ai.machinelearning.native.idl:42]: The come way of doing this is to add a new interface.

ILearningModelSessionOptionsNative1. This way users can QI for the availiblity of this API.

File: winml/api/windows.ai.machinelearning.native.internal.idl [hunk 3]
```diff
@@ -59,6 +59,7 @@ interface ILearningModelEvaluationResultNative : IUnknown
 interface ILearningModelSessionNative : IUnknown
 {
     HRESULT GetIntraOpNumThreads(UINT32* numThreads);
+    HRESULT GetIntraOpThreadSpinning(boolean* allowSpinning);
```
Comment (smk2007) [inline winml/api/windows.ai.machinelearning.native.internal.idl:62]: This should be ILearningModelSessionNative1

File: winml/lib/Api/LearningModelSessionOptions.cpp [hunk 4]
```diff
@@ -42,4 +42,14 @@ STDMETHODIMP LearningModelSessionOptions::SetIntraOpNumThreadsOverride(uint32_t
   telemetry_helper.SetIntraOpNumThreadsOverride(intraOpNumThreads);
   return S_OK;
 }
+
+bool LearningModelSessionOptions::GetIntraOpThreadSpinning() {
+  return allow_thread_spinning_;
```
Comment (smk2007) [inline winml/lib/Api/LearningModelSessionOptions.cpp:47]: No telemetry for the getter?

File: winml/adapter/winml_adapter_apis.h (hunk 1 above)
Comment (orilevari) [reply winml/adapter/winml_adapter_apis.h:59]: As opposed to std::string? Not sure, I remembered the iengine API surface is supposed to stay ABI safe so I just used char ** for all the different levels for consistency. Any reason not to?

File: winml/api/windows.ai.machinelearning.native.idl [hunk 5]
```diff
@@ -35,10 +35,11 @@ interface ILearningModelDeviceFactoryNative : IUnknown
     HRESULT CreateFromD3D12CommandQueue(ID3D12CommandQueue * value, [out] IUnknown ** result);
 };
 
-[uuid(c71e953f-37b4-4564-8658-d8396866db0d), object, local]
+[uuid(5da37a26-0526-414b-91e4-2a0fa3ddba40), object, local]
```
[the remark this answers is not in the document]
Comment (orilevari) [reply winml/api/windows.ai.machinelearning.native.idl:38]: Oh okay, so its okay to leave it the same but add a new api to the block?

File: winml/lib/Api/LearningModelSessionOptions.cpp (hunk 4 above)
Comment (orilevari) [reply winml/lib/Api/LearningModelSessionOptions.cpp:47]: its not exposed in our idl, so its only ever used internally. Its the same pattern for GetIntraOpNumThreads.

File: winml/api/windows.ai.machinelearning.native.idl (hunk 2 above)
Comment (orilevari) [reply winml/api/windows.ai.machinelearning.native.idl:42]: ah I see.

File: winml/adapter/winml_adapter_apis.h (hunk 1 above)
Comment (smk2007) [reply winml/adapter/winml_adapter_apis.h:59]: why not bool?

File: winml/api/windows.ai.machinelearning.native.internal.idl (hunk 3 above)
Comment (orilevari) [reply winml/api/windows.ai.machinelearning.native.internal.idl:67]: fixed

**Review by smk2007 — APPROVED** [review]

**Pull request merged**